[NEXUS-5554] Reorganize NX namespace, allow plugins to specify require.js module/dependencies directly

nabcos opened the pull request:
Use requirejs for most of what was in lib and move to a different namespace.
Introduce support for plugins to specify their js dependencies directly.

nabcos committed:
Reorganize lib to use requirejs

nabcos committed:
Move Mixins to Nexus namespace

jdillon commented:
The namespace took me by surprise a little, but I think after looking this over for a bit that its all good ;-)

+1

nabcos committed:
Fix singleton creation

nabcos committed:
Fix mixin namespace

nabcos committed:
Fix UI loading: Add nx-all.js, order lib/nx/UI loading

Because we use requirejs modules that install stuff into global
namespace (... NX.define ...) and use those things *outside of
an r.js module* (... NX.define defines r.js modules ...) we have
to "layer" our js loading:

1. sonatype-lib.js is mandatory, contains require.js and string utils
2. nx-all.js contains the definitions needed to load the actual
   application code (NX.define et.al.)
3. sonatype-all.js contains the UI app, uses NX namespace

nabcos committed:
Add r.js support for plugins

jdillon commented on nexus-core/src/main/java/org/sonatype/nexus/plugins/rest/RequireJsContributor.java:
```diff
@@ -0,0 +1,72 @@
+
+        private String module;
+
+        public RequireJsContribution( final String module, final List<String> dependencies )
```
If this is expected to be coded up like you have done here:

https://github.com/sonatype/nexus-ssl/commit/a55033a35420b34f60f4ce794fdc32d46d2bfd53

Then you should probably make String... dependencies helper to avoid needing to have the implementation make lists.

Though really I think a builder model here is better, since I'm guessing a lot of plugins will want to have something like you put in nexus-ssl:

```
return new RequireJsContribution( "Nexus/ssl/bootstrap", ImmutableList.of(String.format("static/js/%s-all.js", ARTIFACT_ID)) );
```

And it would be good to simply just have a method which put a dependency on static/js/%s-all.js w/ an artifact id.  Leaving this up to the plugin will invariably leave us with plugins putting sources in the wrong place.  I recommend we hide all of that from the implementation of a require contributor.

Also, keep in mind ExtJS 4.1, whatever new plugin support we add should also be able to translate into ExtJS 4.1 w/o any requirejs.  So I might recommend a better name for this, and cook in support for requirejs and probably add in support to just require a js class (and then translate it to requirejs semantics under the covers).

nabcos committed:
Introduce builder for r.js contrib

nabcos committed:
Revert changes to singletons creation

nabcos committed:
Renaming

nabcos committed:
Generate syntactically correct javascript

jdillon commented:
+1 once its mergable
